server/src/pull/cvr.ts

```diff
@@ -0,0 +1,226 @@
+import type {SearchResult} from '../data.js';
+import type {Executor} from '../pg.js';
+
+export type CVR = {
+  clientGroupID: string;
+  clientVersion: number;
+  order: number;
+};
+
+export async function getCVR(
+  executor: Executor,
+  clientGroupID: string,
+  order: number,
+): Promise<CVR | undefined> {
+  const result = await executor(
+    /*sql*/ `SELECT "client_version" FROM "client_view" WHERE "client_group_id" = $1 AND "version" = $2`,
+    [clientGroupID, order],
+  );
+  if (result.rowCount === 0) {
+    return undefined;
+  }
+  return {
+    clientGroupID,
+    order,
+    clientVersion: result.rows[0].client_version,
+  };
+}
+
+export function putCVR(executor: Executor, cvr: CVR) {
+  return executor(
+    /*sql*/ `INSERT INTO client_view (
+      "client_group_id",
+      "client_version",
+      "version"
+    ) VALUES ($1, $2, $3) ON CONFLICT ("client_group_id", "version") DO UPDATE SET
+      client_version = excluded.client_version
+    `,
+    [cvr.clientGroupID, cvr.clientVersion, cvr.order],
+  );
+}
+
+/**
+ * Find all rows in a table that are not in our CVR table.
+ * - either the id is not present
+ * - or the version is different
+ *
+ * We could do cursoring to speed this along.
+ * I.e., For a given pull sequence, we can keep track of the last id and version.
+ * A "reset pull" brings cursor back to 0 to deal with privacy or query changes.
+ *
+ * But also, if the frontend drives the queries then we'll never actually be fetching
+ * so much as the frontend queries will be reasonably sized. The frontend queries would also
+ * be cursored.
+ *
+ * E.g., `SELECT * FROM issue WHERE modified > x OR (modified = y AND id > z) ORDER BY modified, id LIMIT 100`
+ * Cursored on modified and id.
+ *
+ * This means our compare against the CVR would not be a full table scan of `issue`.
+ */
+export function findUnsentItems(
+  executor: Executor,
+  table: keyof typeof TableOrdinal,
+  clientGroupID: string,
+  order: number,
+  limit: number,
+) {
+  // The query used below is the fastest.
+  // Other query forms that were tried:
+  // 1. 10x slower: SELECT * FROM table WHERE NOT EXISTS (SELECT 1 FROM client_view_entry ...)
+  // 2. 2x slower: SELECT id, version FROM table EXCEPT SELECT entity_id, entity_version FROM client_view_entry ...
+  // 3. SELECT * FROM table LEFT JOIN client_view_entry ...
+  const sql = /*sql*/ `SELECT *
+      FROM "${table}" t
+      WHERE (t."id", t."version") NOT IN (
+        SELECT "entity_id", "entity_version"
+        FROM "client_view_entry"
+        WHERE "client_group_id" = $1
+          AND "client_view_version" <= $2
+          AND "entity" = $3
+      )
+      LIMIT $4;`;
+
+  const params = [clientGroupID, order, TableOrdinal[table], limit];
+  return executor(sql, params);
+}
+
+export const TableOrdinal = {
+  issue: 1,
+  description: 2,
+  comment: 3,
+} as const;
+
+export function findDeletions(
```

Comment:
I'm not completely understanding how deletions work.

Does this correctly handle an entry being put, then deleted, then later being put again?

Reply:
Delete does currently miss this case.

Reply:
No it's not expected. We want people to be able to completely delete rows from the primary table. So their row-version will reset.

But that should be fine. `findUnsentItems` is only comparing versions by equality. It doesn't need them to be ordered. Right?

Reply:
Yeah, that's true for `findUnsentItems` and can do the same for delete + re-insert.

I think I see where the disconnect between @grgbkr and I (on https://github.com/rocicorp/repliear/pull/158#discussion_r1414833422) is coming from now that I'm thinking a bit more about delete.